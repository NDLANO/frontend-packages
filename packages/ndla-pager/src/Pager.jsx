/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React from 'react';
import PropTypes from 'prop-types';
import elementType from 'react-prop-types/lib/elementType';
import { Link } from 'react-router-dom';
import styled from '@emotion/styled';
import { css } from '@emotion/core';
import rgba from 'polished/lib/color/rgba';
import { colors } from '@ndla/core';

import SafeLink from './SafeLink'; // TODO: find a solution for sharing common components
import { stepNumbers } from './pagerHelpers';

const createQueryString = obj =>
  Object.keys(obj)
    .map(key => `${key}=${obj[key]}`)
    .join('&');

const pageItemActiveStyle = css`
  border-top: 3px solid ${colors.brand.primary};
  background-color: ${rgba(colors.brand.lighter, 0.5)};
`;

const pageItemStyle = css`
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid ${colors.background.dark};
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  text-decoration: none;
  box-shadow: none;
  color: ${colors.brand.primary};
  cursor: pointer;
  &:hover {
    ${pageItemActiveStyle};
  }
`;

export const PageItem = ({
  children,
  page,
  query: currentQuery,
  pathname,
  onClick,
  pageItemComponentClass: Component,
}) => {
  const query = { ...currentQuery, page };
  const linkToPage = {
    pathname,
    search: createQueryString(query),
  };

  const handleClick = () => onClick(query);

  if (Component === SafeLink || Component === Link) {
    return (
      <SafeLink css={pageItemStyle} onClick={handleClick} to={linkToPage}>
        {children}
      </SafeLink>
    );
  }
  return (
    <Component css={pageItemStyle} onClick={handleClick}>
      {children}
    </Component>
  );
};

PageItem.propTypes = {
  pageItemComponentClass: elementType.isRequired,
  children: PropTypes.node.isRequired,
  page: PropTypes.number.isRequired,
  query: PropTypes.object.isRequired, // eslint-disable-line
  pathname: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};

const StyledPager = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-top: 1em;
`;

export default function Pager(props) {
  const { page, lastPage, ...rest } = props;

  const steps = stepNumbers(page, lastPage);

  const PageItems = steps.map(n => {
    if (n === page) {
      return (
        <span key={n} css={[pageItemStyle, pageItemActiveStyle]}>
          {n}
        </span>
      );
    }
    return (
      <PageItem key={n} page={n} {...rest}>
        {n}
      </PageItem>
    );
  });

  const prevPageItem =
    steps[0] < page ? (
      <PageItem page={page - 1} {...rest}>
        {' '}
        {'<'}{' '}
      </PageItem>
    ) : null;
  const nextPageItem =
    page < lastPage ? (
      <PageItem page={page + 1} {...rest}>
        {' '}
        {'>'}{' '}
      </PageItem>
    ) : null;

  return (
    <StyledPager>
      {prevPageItem}
      {PageItems}
      {nextPageItem}
    </StyledPager>
  );
}

Pager.propTypes = {
  page: PropTypes.number.isRequired,
  lastPage: PropTypes.number.isRequired,
  pathname: PropTypes.string,
  query: PropTypes.object, // eslint-disable-line
  onClick: PropTypes.func,
  pageItemComponentClass: elementType,
};

Pager.defaultProps = {
  onClick: () => {},
  pageItemComponentClass: SafeLink,
  query: {},
  pathname: '',
};
