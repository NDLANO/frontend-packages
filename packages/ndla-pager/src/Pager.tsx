/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

/** @jsxImportSource @emotion/react */
import { ElementType, ReactNode } from "react";
import { css } from "@emotion/react";
import styled from "@emotion/styled";
import { colors } from "@ndla/core";
import { SafeLink } from "@ndla/safelink";
import { stepNumbers } from "./pagerHelpers";

const createQueryString = (obj: Query) =>
  Object.entries(obj)
    .filter(([_, value]) => value !== undefined && value !== "")
    .map(([key, value]) => `${key}=${value}`)
    .join("&");

type ColorTheme = "primary" | "lighter";

const pageItemActiveStyle = css`
  border-top: 3px solid ${colors.brand.primary};
  background-color: "rgba(222,235,246,0.5)";
`;

const pageItemStyle = (small = false, color: ColorTheme) => css`
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid ${color === "primary" ? colors.background.dark : colors.white};
  padding: ${small ? "0.2em 0.6em" : "1em 1.45em"};
  margin: 1em 0.3em 0;
  text-decoration: none;
  box-shadow: none;
  color: ${colors.brand.primary};
  cursor: pointer;
  &:hover {
    ${pageItemActiveStyle};
  }
`;

const StyledSafeLink = styled(SafeLink)`
  background: transparent;
  border: 0;
  border-top: 3px solid ${colors.white};
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  text-decoration: none;
  box-shadow: none;
  color: ${colors.brand.primary};
  cursor: pointer;
  &[data-color="primary"] {
    border-top: 3px solid ${colors.background.dark};
  }
  &[data-small="true"] {
    padding: 0.2em 0.6em;
  }
  &:hover {
    ${pageItemActiveStyle};
  }
`;

type Query = Record<string, any>;

interface PageItemProps<T extends Query> {
  pageItemComponentClass: ElementType;
  children: ReactNode;
  page: number;
  query: T;
  pathname: string;
  onClick: (query: T & { page: number }) => void;
  type?: string;
  small: boolean;
  colorTheme: ColorTheme;
}

export const PageItem = <T extends Query>({
  children,
  page,
  query: currentQuery,
  pathname,
  onClick,
  pageItemComponentClass: Component,
  type,
  small,
  colorTheme,
}: PageItemProps<T>) => {
  const query = { ...currentQuery, page };
  const linkToPage = {
    pathname,
    search: createQueryString(query),
  };

  const handleClick = () => onClick(query);

  if (Component === SafeLink) {
    return (
      <StyledSafeLink data-color={colorTheme} data-small={small} onClick={handleClick} to={linkToPage}>
        {children}
      </StyledSafeLink>
    );
  }
  return (
    <Component css={pageItemStyle(small, colorTheme)} onClick={handleClick} type={type}>
      {children}
    </Component>
  );
};

const StyledPager = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: center;
  margin-top: 1em;
`;

interface Props<T extends Query> {
  page: number;
  lastPage: number;
  pathname?: string;
  query: T;
  onClick?: (query: T & { page: number }) => void;
  pageItemComponentClass?: ElementType;
  small?: boolean;
  colorTheme?: ColorTheme;
}

const Pager = <T extends Query>({
  page,
  lastPage,
  onClick = (_) => {},
  pageItemComponentClass = SafeLink,
  query,
  pathname = "",
  small = false,
  colorTheme = "primary",
}: Props<T>) => {
  const steps = stepNumbers(page, lastPage);

  const rest = {
    onClick,
    pageItemComponentClass,
    query,
    pathname,
    small,
    colorTheme,
  };
  const PageItems = steps.map((n) => {
    if (n === page) {
      return (
        <span key={n} aria-current="step" css={[pageItemStyle(small, colorTheme), pageItemActiveStyle]}>
          {n}
        </span>
      );
    }
    return (
      <PageItem key={n} page={n} {...rest}>
        {n}
      </PageItem>
    );
  });

  const prevPageItem =
    steps[0] < page ? (
      <PageItem page={page - 1} {...rest}>
        {"<"}
      </PageItem>
    ) : null;
  const nextPageItem =
    page < lastPage ? (
      <PageItem page={page + 1} {...rest}>
        {">"}
      </PageItem>
    ) : null;

  return (
    <StyledPager data-testid="pager">
      {prevPageItem}
      {PageItems}
      {nextPageItem}
    </StyledPager>
  );
};

export default Pager;
