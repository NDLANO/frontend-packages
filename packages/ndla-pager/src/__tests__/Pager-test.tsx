/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { ReactNode } from "react";
import { MemoryRouter } from "react-router-dom";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";

import Pager from "../Pager";

interface PagerTestParams {
  setup: { page: number; lastPage: number };
  expected: (string | number)[];
}

interface WrapperProps {
  children?: ReactNode;
}

const wrapper = ({ children }: WrapperProps) => <MemoryRouter>{children}</MemoryRouter>;

const pagerTest = ({ setup, expected }: PagerTestParams) => {
  test(`component/LinkPager page ${setup.page}/${setup.lastPage}`, async () => {
    const path = "somepath";
    const prev = setup.page - 1;
    const next = setup.page + 1;
    const { findByTestId } = render(<Pager pathname={path} query={{}} {...setup} />, {
      wrapper,
    });
    const pager = await findByTestId("pager");
    expect(pager.children.length).toBe(expected.length);

    expected.forEach((value, i) => {
      const step = pager.children[i];
      switch (value) {
        case "current":
          expect(step).toHaveTextContent(setup.page.toString());
          expect(step).not.toHaveAttribute("href");
          expect(step).toHaveAttribute("aria-current", "step");
          break;
        case "back":
          expect(step).toHaveTextContent("<");
          expect(step).toHaveAttribute("href", `/${path}?page=${prev}`);
          break;
        case "forward":
          expect(step).toHaveTextContent(">");
          expect(step).toHaveAttribute("href", `/${path}?page=${next}`);
          break;
        default:
          expect(step).toHaveTextContent(value.toString());
          expect(step).toHaveAttribute("href", `/${path}?page=${value}`);
      }
    });
  });
};

pagerTest({ setup: { page: 1, lastPage: 1 }, expected: ["current"] });
pagerTest({
  setup: { page: 3, lastPage: 5 },
  expected: ["back", 1, 2, "current", 4, 5, "forward"],
});
pagerTest({
  setup: { page: 1, lastPage: 5 },
  expected: ["current", 2, 3, 4, 5, "forward"],
});
pagerTest({
  setup: { page: 19, lastPage: 19 },
  expected: ["back", 15, 16, 17, 18, "current"],
});
pagerTest({
  setup: { page: 4, lastPage: 10 },
  expected: ["back", 2, 3, "current", 5, 6, "forward"],
});
