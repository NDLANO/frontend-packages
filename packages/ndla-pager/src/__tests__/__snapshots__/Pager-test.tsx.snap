// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component/LinkPager page 1/1 1`] = `
.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 1em;
}

.emotion-0 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-0:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

<div>
  <div
    class="emotion-1"
  >
    <span
      class="emotion-0"
    >
      1
    </span>
  </div>
</div>
`;

exports[`component/LinkPager page 1/5 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 1em;
}

.emotion-0 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-0:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-1 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
}

.emotion-1:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

<div>
  <div
    class="emotion-6"
  >
    <span
      class="emotion-0"
    >
      1
    </span>
    <a
      class="emotion-1"
      href="somepath?page=2"
      tabindex="0"
    >
      2
    </a>
    <a
      class="emotion-1"
      href="somepath?page=3"
      tabindex="0"
    >
      3
    </a>
    <a
      class="emotion-1"
      href="somepath?page=4"
      tabindex="0"
    >
      4
    </a>
    <a
      class="emotion-1"
      href="somepath?page=5"
      tabindex="0"
    >
      5
    </a>
    <a
      class="emotion-1"
      href="somepath?page=2"
      tabindex="0"
    >
       
      &gt;
       
    </a>
  </div>
</div>
`;

exports[`component/LinkPager page 3/5 1`] = `
.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 1em;
}

.emotion-3 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-3:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-0 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
}

.emotion-0:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

<div>
  <div
    class="emotion-7"
  >
    <a
      class="emotion-0"
      href="somepath?page=2"
      tabindex="0"
    >
       
      &lt;
       
    </a>
    <a
      class="emotion-0"
      href="somepath?page=1"
      tabindex="0"
    >
      1
    </a>
    <a
      class="emotion-0"
      href="somepath?page=2"
      tabindex="0"
    >
      2
    </a>
    <span
      class="emotion-3"
    >
      3
    </span>
    <a
      class="emotion-0"
      href="somepath?page=4"
      tabindex="0"
    >
      4
    </a>
    <a
      class="emotion-0"
      href="somepath?page=5"
      tabindex="0"
    >
      5
    </a>
    <a
      class="emotion-0"
      href="somepath?page=4"
      tabindex="0"
    >
       
      &gt;
       
    </a>
  </div>
</div>
`;

exports[`component/LinkPager page 4/10 1`] = `
.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 1em;
}

.emotion-3 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-3:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-0 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
}

.emotion-0:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

<div>
  <div
    class="emotion-7"
  >
    <a
      class="emotion-0"
      href="somepath?page=3"
      tabindex="0"
    >
       
      &lt;
       
    </a>
    <a
      class="emotion-0"
      href="somepath?page=2"
      tabindex="0"
    >
      2
    </a>
    <a
      class="emotion-0"
      href="somepath?page=3"
      tabindex="0"
    >
      3
    </a>
    <span
      class="emotion-3"
    >
      4
    </span>
    <a
      class="emotion-0"
      href="somepath?page=5"
      tabindex="0"
    >
      5
    </a>
    <a
      class="emotion-0"
      href="somepath?page=6"
      tabindex="0"
    >
      6
    </a>
    <a
      class="emotion-0"
      href="somepath?page=5"
      tabindex="0"
    >
       
      &gt;
       
    </a>
  </div>
</div>
`;

exports[`component/LinkPager page 19/19 1`] = `
.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin-top: 1em;
}

.emotion-5 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-5:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

.emotion-0 {
  display: inline;
  background: transparent;
  border: 0;
  border-top: 3px solid #e6e6e6;
  padding: 1em 1.45em;
  margin: 1em 0.3em 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  box-shadow: none;
  color: #20588f;
  cursor: pointer;
}

.emotion-0:hover {
  border-top: 3px solid #20588f;
  background-color: rgba(222,235,246,0.5);
}

<div>
  <div
    class="emotion-6"
  >
    <a
      class="emotion-0"
      href="somepath?page=18"
      tabindex="0"
    >
       
      &lt;
       
    </a>
    <a
      class="emotion-0"
      href="somepath?page=15"
      tabindex="0"
    >
      15
    </a>
    <a
      class="emotion-0"
      href="somepath?page=16"
      tabindex="0"
    >
      16
    </a>
    <a
      class="emotion-0"
      href="somepath?page=17"
      tabindex="0"
    >
      17
    </a>
    <a
      class="emotion-0"
      href="somepath?page=18"
      tabindex="0"
    >
      18
    </a>
    <span
      class="emotion-5"
    >
      19
    </span>
  </div>
</div>
`;
